### ISSUE TITLE
[SCM-791] GitExe Provider does not work properly in Cygwin

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-scm#1019]
**[Irineu Ruiz](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=irineu.ruiz)** opened **[SCM-791](https://issues.apache.org/jira/browse/SCM-791?redirect=false)** and commented

The tests of the maven-scm project fails in CygWin and passed in command prompt, as you can see in the attachment.

Also, using Cygwin under Windows 8, we can't execute a mvn release:prepare properly, because pom.xml is not correct commited in tag and HEAD of repository.
The maven gives a false positive of Build Sucess, but the tag is incorret, with the SNAPSHOT termination in version.

---

**Affects:** 1.9.2

**Attachments:**
- [bad-01ba0c0.txt](https://issues.apache.org/jira/secure/attachment/12935271/bad-01ba0c0.txt) (_18.12 kB_)
- [good-SCM-791.txt](https://issues.apache.org/jira/secure/attachment/12935272/good-SCM-791.txt) (_20.35 kB_)
- [test_1.9.2_Cygwin.log](https://issues.apache.org/jira/secure/attachment/12718724/test_1.9.2_Cygwin.log) (_95.88 kB_)
- [test_1.9.2_prompt.log](https://issues.apache.org/jira/secure/attachment/12718634/test_1.9.2_prompt.log) (_270.96 kB_)
- [test_HEAD_Cygwin.log](https://issues.apache.org/jira/secure/attachment/12718576/test_HEAD_Cygwin.log) (_97.99 kB_)
- [test_HEAD_prompt.log](https://issues.apache.org/jira/secure/attachment/12718635/test_HEAD_prompt.log) (_314.10 kB_)

**Issue Links:**
- [SCM-815](https://issues.apache.org/jira/browse/SCM-815) GitExe Provider does not commit files in Cygwin

**Remote Links:**
- [GitHub Pull Request #77
](https://github.com/apache/maven-scm/pull/77)

0 votes, 7 watchers

**Comment by jira-importer** [comment]
**[Dan Tran](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=danttran)** commented

looks like cygwin git does not like windows path.

BTW, I dont think svn under cygwin ever work with release plugin

**Comment by jira-importer** [comment]
**[Christian Galsterer](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=christiangalsterer)** commented

I think the problem was introduced in 1.9.1 especially here https://github.com/apache/maven-scm/blob/maven-scm-1.9.1/maven-scm-providers/maven-scm-providers-git/maven-scm-provider-gitexe/src/main/java/org/apache/maven/scm/provider/git/gitexe/command/status/GitStatusConsumer.java#L214

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Just downloaded cygwin and ran the tests. Cannot reproduce. Please comment.

**Comment by jira-importer** [comment]
**[Ilya Basin](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=basinilya)** commented

Did you forget to add cygwin git to path? See attached [^bad-01ba0c0.txt].

**Comment by jira-importer** [comment]
**[Ilya Basin](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=basinilya)** commented

BTW to make my patch work you need to make Cygwin git understand some valid Git-For-Windows URLs:

```java
git config --global   url.file:///cygdrive/c.insteadOf   file:///C: file:///C:
git config --global   url.file:///cygdrive/c.insteadOf   file://C: file://C:
git config --global   url.file:////.insteadOf            file://///
```

This is unavoidable trade-off as I explained in my [mail](http://mail-archives.apache.org/mod_mbox/maven-dev/201807.mbox/%3C404008c7-7332-d599-f2a6-bd91dfc8fcd8%40gmail.com%3E)

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Alright, added the Git config and can reproduce now. I have found several other bugs where a `File` is passed for a URL. I've fixed those too.

Please check my branch `path-fiddle`. Currently, `GitExeChangeLogTckTest` fails because `GitChangeLogCommand` does this: `cl.createArg().setFile( workingDirectory )`. This change is valid, but has to be relative to the repo root to work properly with Cygwin. Also testcases need to be changed.

**Comment by jira-importer** [comment]
**[Ilya Basin](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=basinilya)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)
It helps to modify GitChangeLogCommand like this: cl.createArg().setValue( "." );
We do "cd workdir" to run shell commands on Linux, don't we?
The expected values in GitChangeLogCommandTest should be updated accordingly.

I see you have removed getAbsolutePath from GitScmTestUtils.getScmUrl. That's not right. Users should be able to do something like this: "scm:git:file://${basedir}" and "${basedir}" may be like "C:\something" or "\\unc\path". This is why I replace the slashes so late in GitScmProviderRepository.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Thanks for the review. I have updated the branch. Please have a look GitExe now fully passed, JGit makes some troube...investigating.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

JGit works flawlessly as soon as you disable the URL mappings for Cygwin.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Fixed with [dc9b283269ac67ebbaae1d285f58d591b0f28573](https://gitbox.apache.org/repos/asf?p=maven-scm.git;a=commit;h=dc9b283269ac67ebbaae1d285f58d591b0f28573).
